src/coreclr/src/jit/hwintrinsicxarch.cpp

```diff
@@ -528,6 +528,118 @@ GenTree* Compiler::impBaseIntrinsic(NamedIntrinsic        intrinsic,
             break;
         }
 
+        case NI_Vector128_AsVector:
+        {
+            assert(sig->numArgs == 1);
+
+            if (getSIMDVectorRegisterByteLength() == YMM_REGSIZE_BYTES)
+            {
+                // Vector<T> is 32-bytes, so we should treat this as a call to Vector128.ToVector256
+                return impBaseIntrinsic(NI_Vector128_ToVector256, clsHnd, method, sig, mustExpand);
+            }
+
+            assert(getSIMDVectorRegisterByteLength() == XMM_REGSIZE_BYTES);
+
+            // We fold away the cast here, as it only exists to satisfy
+            // the type system. It is safe to do this here since the retNode type
+            // and the signature return type are both the same TYP_SIMD.
+
+            retNode = impSIMDPopStack(retType);
+            SetOpLclRelatedToSIMDIntrinsic(retNode);
+            assert(retNode->gtType == getSIMDTypeForSize(getSIMDTypeSizeInBytes(sig->retTypeSigClass)));
+
+            break;
+        }
+
+        case NI_Vector128_AsVector2:
+        case NI_Vector128_AsVector3:
+        case NI_Vector128_AsVector4:
```

Comment:
`AsVector4` is the simplest, we have a `TYP_SIMD16` and are going to a `TYP_SIMD16`

`AsVector2` and `AsVector3` should also be simple. We have a `TYP_SIMD16` and we want a `TYP_SIMD8` or `TYP_SIMD12`. If they are already in register, there is no change needed. If they are in memory, we have 16-bytes and are correctly aligned, so it should likewise be safe to just access.

Reply:
> If they are already in register, there is no change needed

I'm not so sure. Some SIMD8/12 operations intentionally zero out the upper unused elements (SIMDIntrinsicDiv and maybe others). Either that zeroing is unnecessary or AsVector2/3 should also zero out.

Reply:
I believe that's just to avoid the penalty around `x / 0` though rather than a functional requirement.

We may want to have both options a "simple, but correct" API. That is, one which ensures that the upper bits are correctly zeroed
and a "fast, but unsafe" API. That is, one which leaves clearing/setting the upper bits to the user; that way the conversion is as low as possible (generally free).

Reply:
@CarolEidt, do you have an opinion here?

Should I, when converting `Vector128<float>` to `Vector2/3`, explicitly zero the untracked `z`/`w` fields or should that responsibility be left up to the consumer if they feel it is important?

As best as I can tell, it doesn't matter from a functional perspective and at worst could cause a minor perf hit if those fields end up containing a subnormal value or NaN.

Reply:
Wouldn't the `Dot` methods break if the top elements aren't zeroed?

Reply:
No, `dpps` allows you to specify which elements are consumed and which are ignored as part of the computation.

Reply:
Looks like we might need to update the codegen for `Vector2.Dot` however. It currently generates the same as `Vector4` using `0xF1` as the control word and it should be `0x3F` or `0x31`

Reply:
Yeah, just saw that.  The SSE2 fallback for `Vector2` is also using the `Vector4` code from what I can tell

Reply:
Sounds like some room for improvement there then. I'll take a look at a few other cases and will log bugs to start fixing them.

Reply:
I reverted the change treating `AsVector2`/`AsVector3` as intrinsic for right now since it requires some fixes to the SIMD code and since it looks like it requires a few fixes elsewhere in the JIT as well.